### PR TITLE
Memory leak fixes

### PR DESCRIPTION
**Pull request by hbmartin**
This PR is the result of a memory leak investigation started in https://github.com/apollographql/apollo-kotlin/issues/5262
The cause seems to be the KN GC incorrectly not releasing strings in a CFString -> NSString -> (Kotlin) String allocation
This PR attempts to fix this in 3 ways: 

1. Allocate a single empty string constant as a fallback for `sqlite3_column_text` instead of allocating a new empty string on each call.
2. Wrap the `bytesToString` call for `appleMain` in an `autoreleasepool` block.
3. Upgrading to Kotlin 1.9.20 to take advantage of the `custom` allocator.

The Kotlin upgrade entailed other work:

1. Removing support for mingwX86
2. Configuring additional opt-in flags
3. Disabling the new default hierarchy
4. Migrating `AtomicInt` references

Look forward to feedback and discussion here!

### CODE REVIEW

File: sqliter-driver/src/nativeCommonMain/kotlin/co/touchlab/sqliter/interop/Types.kt [hunk 1]
```diff
@@ -3,5 +3,5 @@ package co.touchlab.sqliter.interop
 import cnames.structs.sqlite3_stmt
 import kotlinx.cinterop.CPointer
 
-typealias SqliteDatabasePointer = CPointer<co.touchlab.sqliter.sqlite3.sqlite3>
+typealias SqliteDatabasePointer = CPointer<cnames.structs.sqlite3>
```
Comment (hbmartin) [inline sqliter-driver/src/nativeCommonMain/kotlin/co/touchlab/sqliter/interop/Types.kt:6]: this was a surprising change to make, it seems something in 1.9.20 changed cinterop packaging?

Comment (JakeWharton) [reply sqliter-driver/src/nativeCommonMain/kotlin/co/touchlab/sqliter/interop/Types.kt:6]: Probably a regression of https://youtrack.jetbrains.com/issue/KT-49034

You can work around it by defining the struct manually in your .def like this: https://github.com/cashapp/zipline/blob/ce04291109a5f2e4487bc762bc14a681013836c9/zipline/src/nativeInterop/cinterop/quickjs.def#L7-L10

Comment (hbmartin) [reply sqliter-driver/src/nativeCommonMain/kotlin/co/touchlab/sqliter/interop/Types.kt:6]: I'm probably holding it wrong, but I tried a few variations e.g. `typedef struct sqlite3 {} sqlite3;` , `typedef struct sqlite3;` etc and all seemed to have the same issue.

Comment (JakeWharton) [reply sqliter-driver/src/nativeCommonMain/kotlin/co/touchlab/sqliter/interop/Types.kt:6]: I'll try on Zipline on Tuesday and see what happens.

**Comment by kpgalligan** [comment]
Assume the autorelease is cleaning up the `NSString` returned from `NSString.create`. That's a surprise. Will take a deeper look tomorrow/Tuesday and push out a build.

File: sqliter-driver/src/linuxX64Main/kotlin/co/touchlab/sqliter/internal/File.kt [hunk 2]
```diff
@@ -431,7 +431,7 @@ internal actual class File(dirPath: String? = null, name: String) {
     }
 
     private fun mkdirImpl(filePath: String): Boolean {
-        return mkdir(filePath, S_IRWXU) == F_OK
+        return mkdir(filePath, S_IRWXU.toUInt()) == F_OK
```
Comment (hbmartin) [inline sqliter-driver/src/linuxX64Main/kotlin/co/touchlab/sqliter/internal/File.kt:434]: `S_IRWXU` is an `Int` but `mkdir` signature is now `@kotlinx.cinterop.internal.CCall public external fun mkdir(@kotlinx.cinterop.internal.CCall.CString __path: kotlin.String?, __mode: platform.posix.__mode_t /* = kotlin.UInt */): kotlin.Int { /* compiled code */ }`

**Comment by hbmartin** [comment]
I did not anticipate fighting with junit to be the hard part of this... still trying to solve

**Comment by hbmartin** [comment]
I took a stab at reading the binary test outputs https://gist.github.com/hbmartin/a2af9046f0d05c1b6a6d64267136d933
It seems that recording the entire sqlite logging output during tests was causing some overflows in those outputs.
My fix was to create a `NoneLogger` and use that for all test cases. 🤷 

I also noticed that a couple of tests were calling `co.touchlab.sqliter.internal.File.exists` property as a function and fixed that. Seems odd that this ever worked - my wild guess is that the complier may have allowed this since the `actual`s were implemented as backing properties.

**Pull request merged**